<a href="https://colab.research.google.com/github/dnevo/Practicum/blob/master/S07_Introduction_to_ML_%E2%80%93_customer_package_plan_for_Mobile_carrier.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Review

Hi Doron. My name is Soslan. I'm reviewing your work. I've added all my comments to new cells with the title "Review". My apologies for the delay in the review. We will be faster next time :)

```diff
+ If you did something great I'm using green color for my comment
- If the topic requires some extra work so I can accept it then the color will be red.
```

Your project is of good quality. All the checkpoints were done correctly. You did your project in most compact way. Thank you for your work and good luck in future learning.

---

## The project
Mobile carrier Megaline has found out that many of their subscribers use legacy plans. They want to develop a model that would analyze subscribers' behavior and recommend one of Megaline's newer plans: Smart or Ultra.

You have access to behavior data about subscribers who have already switched to the new plans (from the project for the Statistical Data Analysis course). For this classification task, you need to develop a model that will pick the right plan. Since you’ve already performed the data preprocessing step, you can move straight to creating the model.

Develop a model with the highest possible accuracy. In this project, the threshold for accuracy is 0.75. Check the accuracy using the test dataset.

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier
pd.set_option('display.width', 200)
pd.set_option('display.max_columns', None)
pd.options.display.float_format = '{:11,.2f}'.format

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/dnevo/Practicum/master/datasets/users_behavior.csv',index_col=0)
df.head()

minutes    messages     mb_used  is_ultra
calls                                               
 40.00      311.90       83.00   19,915.42         0
 85.00      516.75       56.00   22,696.96         0
 77.00      467.66       86.00   21,060.45         0
106.00      745.53       81.00    8,437.39         1
 66.00      418.74        1.00   14,502.75         0

## Review

OK. It is always good idea too check data for missing values etc. 

---

### Data set Split
The dataset is splitted into 3 groups according to the common practice: train 60%, Validation 20%, test 20%.

Since the train_test_split method is designed to split into only 2 groups, we are calling it twice in order to split to three.

In [3]:
df_train, df_temp = train_test_split(df, test_size=0.4, random_state=12345)
df_valid, df_test = train_test_split(df_temp, test_size=0.5, random_state=12345)

features_train = df_train.drop(['is_ultra'], axis=1)
target_train = df_train['is_ultra']
features_valid = df_valid.drop(['is_ultra'], axis=1)
target_valid = df_valid['is_ultra']
features_test = df_test.drop(['is_ultra'], axis=1)
target_test = df_test['is_ultra']

### Random Forest: best result with n_estimator = 18 (0.782)

In [4]:
for n_estim in range(1,20):
    model = RandomForestClassifier(random_state=12345, n_estimators=n_estim)
    model.fit(features_train, target_train)
    predictions_valid = model.predict(features_valid)
    print("RandomForest n_estimator =", n_estim, ": ", end='')
    print(accuracy_score(target_valid, predictions_valid))

RandomForest n_estimator = 1 : 0.6749611197511665
RandomForest n_estimator = 2 : 0.7356143079315708
RandomForest n_estimator = 3 : 0.7231726283048211
RandomForest n_estimator = 4 : 0.7465007776049767
RandomForest n_estimator = 5 : 0.7465007776049767
RandomForest n_estimator = 6 : 0.7573872472783826
RandomForest n_estimator = 7 : 0.749611197511664
RandomForest n_estimator = 8 : 0.7527216174183515
RandomForest n_estimator = 9 : 0.7511664074650077
RandomForest n_estimator = 10 : 0.7589424572317263
RandomForest n_estimator = 11 : 0.7573872472783826
RandomForest n_estimator = 12 : 0.7651632970451011
RandomForest n_estimator = 13 : 0.7573872472783826
RandomForest n_estimator = 14 : 0.7698289269051322
RandomForest n_estimator = 15 : 0.776049766718507
RandomForest n_estimator = 16 : 0.776049766718507
RandomForest n_estimator = 17 : 0.776049766718507
RandomForest n_estimator = 18 : 0.7822706065318819
RandomForest n_estimator = 19 : 0.7807153965785381


### Logistic Regression: accuracy on validation set: 0.707

In [5]:
model = LogisticRegression(random_state=12345)
model.fit(features_train, target_train)
predictions_valid = model.predict(features_valid)
print("LogisticRegression:", accuracy_score(target_valid, predictions_valid))

LogisticRegression: 0.7076205287713841


### Decision Tree: best result is with max_depth = 4 (0.788)

In [6]:
for depth in range(1,6):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    model.fit(features_train, target_train)
    predictions_valid = model.predict(features_valid)
    print("DecisionTree max_depth =", depth, ": ", end='')
    print(accuracy_score(target_valid, predictions_valid))

DecisionTree max_depth = 1 : 0.7542768273716952
DecisionTree max_depth = 2 : 0.7822706065318819
DecisionTree max_depth = 3 : 0.7853810264385692
DecisionTree max_depth = 4 : 0.7884914463452566
DecisionTree max_depth = 5 : 0.7884914463452566


### Summary of above
We have made a comparison of 3 models: Random Forest, Logistic Regression and Decision Tree. As from the results, Decision Tree (with max_depth = 4) provides best accuracy: 78.8%

In [7]:
model = DecisionTreeClassifier(random_state=12345, max_depth=4)
model.fit(features_train, target_train)
predictions_valid = model.predict(features_valid)
print(accuracy_score(target_valid, predictions_valid))

0.7884914463452566


## Review

```diff
+ Everything is comact and correct. Good job.
```

---

### Conclusion
The last step, is checking accuracy on the test set. The accuracy remains similar 78.1% - which is above the target of 75% that we have set.

In [8]:
predictions_test = model.predict(features_test)
print(accuracy_score(target_test, predictions_test))

0.7807153965785381


## Review

```diff
+ All steps were done correctly
```

---